activerecord/lib/active_record/associations/collection_association.rb

```diff
@@ -274,21 +274,19 @@ def load_target
         target
       end
 
-      def add_to_target(record, skip_callbacks: false, replace: false, &block)
-        replace_on_target(record, skip_callbacks, replace: replace || association_scope.distinct_value, &block)
+      def add_to_target(record, skip_callbacks: false, &block)
+        add_record(record, replace: association_scope.distinct_value, skip_callbacks: skip_callbacks, &block)
```

Comment:
Doesn't this changes the behavior on `build` when `replace:` was being passed as `true`? Now it will use `association_scope.distinct_value` to determine if it should replace or not.

Reply:
If I understand correctly, `replace: true` was added in 50e4aaaf25557fca9cb712b15323b898bbe32665 because the following line ends up calling `replace_on_target` twice:

https://github.com/rails/rails/blob/ec4558ae9f5cc60355f3fd15d4f2f7adccf1c158/activerecord/lib/active_record/associations/collection_association.rb#L119

The 1st time [via `target=`](https://github.com/rails/rails/blob/ec4558ae9f5cc60355f3fd15d4f2f7adccf1c158/activerecord/lib/active_record/associations/collection_association.rb#L290) inside the `build_record` call, and the 2nd time via `add_to_target`.

For the 2nd call, `replace: true` causes the following conditional to compute `index`:

https://github.com/rails/rails/blob/ec4558ae9f5cc60355f3fd15d4f2f7adccf1c158/activerecord/lib/active_record/associations/collection_association.rb#L449-L452

However, a "fallback" was also added in 6f6c441662e73084087da3f564e71f0a174c3ee4 to compute `index` anyway:

https://github.com/rails/rails/blob/ec4558ae9f5cc60355f3fd15d4f2f7adccf1c158/activerecord/lib/active_record/associations/collection_association.rb#L464-L466

In other words, I believe `replace: true` inside `build` is no longer necessary even on `main`.

That said, this PR changes how targets are tracked.  With this PR, [`CollectionAssociation#inversed_from`](https://github.com/rails/rails/blob/ed79453a49d205dec8dd439c760f4fb191594a4e/activerecord/lib/active_record/associations/collection_association.rb#L281-L286) will override [`Association#inversed_from`](https://github.com/rails/rails/blob/ec4558ae9f5cc60355f3fd15d4f2f7adccf1c158/activerecord/lib/active_record/associations/association.rb#L138-L140), so `build_record` inside `build` will call `add_record` with `implicit: true` instead of `target=`.

Thus the 2nd call to `add_to_target` / `add_record` will find the record in `@implicit_target_records` and replace it at `index`:

https://github.com/rails/rails/blob/ed79453a49d205dec8dd439c760f4fb191594a4e/activerecord/lib/active_record/associations/collection_association.rb#L458-L462

It has been a long time since I looked at this code though, so I could be missing something.  A while ago I asked Eileen to test this branch against the Shopify codebase, and she said there were errors but the cause wasn't clear.  Unfortunately, I was not able to dig further.

Reply:
Yeah. I just tried this PR again and there were some errors. Some of them is because we were calling `add_to_target` directly with `replace: true`. The others were is not clear. But points to me that some behavior is changing in this refactoring, so I was trying to see visually what would be the low hanging fruits. I can dig deeper.